xmtp_mls/src/groups/device_sync.rs

```diff
@@ -399,540 +358,514 @@ where
-            DeviceSyncKeyType::Aes256Gcm(key) => DeviceSyncKeyTypeProto {
-                key: Some(EncKeyProto::Aes256Gcm(key.to_vec())),
-            },
-        }
+        Ok(sync_group)
     }
-}
 
-impl TryFrom<DeviceSyncKeyTypeProto> for DeviceSyncKeyType {
-    type Error = DeviceSyncError;
-    fn try_from(key: DeviceSyncKeyTypeProto) -> Result<Self, Self::Error> {
-        let DeviceSyncKeyTypeProto { key } = key;
-        match key {
-            Some(k) => {
-                let EncKeyProto::Aes256Gcm(key) = k;
-                match key.try_into() {
-                    Ok(array) => Ok(DeviceSyncKeyType::Aes256Gcm(array)),
-                    Err(_) => Err(DeviceSyncError::Conversion),
+    /// This should be triggered when a new sync group appears,
+    /// indicating the presence of a new installation.
+    pub async fn add_new_installation_to_groups(&self) -> Result<(), DeviceSyncError> {
+        let provider = self.mls_provider()?;
+        let groups = self.find_groups(GroupQueryArgs::default())?;
+
+        // Add the new installation to groups in batches
+        for chunk in groups.chunks(20) {
+            let mut add_futs = vec![];
+            for group in chunk {
+                add_futs.push(group.add_missing_installations(&provider));
+            }
+            let results = join_all(add_futs).await;
+            for result in results {
+                if let Err(err) = result {
+                    tracing::warn!("Unable to add new installation to group. {err:?}");
                 }
             }
-            None => Err(DeviceSyncError::Conversion),
         }
+
+        Ok(())
     }
 }
 
-pub(super) fn new_request_id() -> String {
-    xmtp_common::rand_string::<ENC_KEY_SIZE>()
+fn default_backup_options() -> BackupOptions {
+    BackupOptions {
+        elements: vec![
+            BackupElementSelection::Messages as i32,
+            BackupElementSelection::Consent as i32,
+        ],
+        ..Default::default()
+    }
 }
 
-pub(super) fn generate_nonce() -> [u8; NONCE_SIZE] {
-    xmtp_common::rand_array::<NONCE_SIZE>()
+// These are the messages that get sent out to the sync group
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+#[repr(i32)]
+pub enum DeviceSyncContent {
+    Request(DeviceSyncRequestProto) = 0,
+    Payload(DeviceSyncReplyProto) = 1,
+    Acknowledge(AcknowledgeKind) = 2,
+    PreferenceUpdates(Vec<UserPreferenceUpdate>) = 3,
 }
-
-pub(super) fn new_pin() -> String {
-    let mut rng = crypto_utils::rng();
-    let pin: u32 = rng.gen_range(0..10000);
-    format!("{:04}", pin)
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub enum AcknowledgeKind {
+    SyncGroupPresence,
+    Request { request_id: String },
 }
 
-fn encrypt_syncables(
-    syncables: &[Vec<Syncable>],
-) -> Result<(Vec<u8>, DeviceSyncKeyType), DeviceSyncError> {
-    let enc_key = DeviceSyncKeyType::new_aes_256_gcm_key();
-    encrypt_syncables_with_key(syncables, enc_key)
+pub trait ZipContent<A, B> {
```

Comment:
how come we can't use [zip](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.zip)

Reply:
I don't think using zip would be an improvement. I could create a new iter that borrowed the data structure we're zipping with, but for what benefit? It's possible, but it feels like we'd be using zip for the sake of using zip. I can rename the trait to be `IterWithContent`